src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs

```diff
@@ -39,18 +40,94 @@ public override bool TryRead(Type valueType, ref Utf8JsonReader reader, out TVal
                 return false;
             }
 
-            if (s_isUint64)
-            {
-                if (reader.TryGetUInt64(out ulong ulongValue))
-                {
-                    value = (TValue)Enum.ToObject(valueType, ulongValue);
-                    return true;
-                }
-            }
-            else if (reader.TryGetInt64(out long longValue))
+            // When utf8reader/writer will support all primitive types we should remove custom bound checks
+            // https://github.com/dotnet/corefx/issues/36125
+
+            switch (s_enumTypeCode)
             {
-                value = (TValue)Enum.ToObject(valueType, longValue);
-                return true;
+                case TypeCode.Empty:
+                case TypeCode.Object:
+                case TypeCode.DBNull:
+                case TypeCode.Boolean:
+                case TypeCode.Char:
+                case TypeCode.Decimal:
+                case TypeCode.Double:
+                case TypeCode.DateTime:
+                case TypeCode.String:
+                case TypeCode.Single:
+                    break;
+                case TypeCode.SByte:
+                    {
+                        if (reader.TryGetInt32(out int byte8) && JsonHelpers.IsInRangeInclusive(byte8, sbyte.MinValue, sbyte.MaxValue))
+                        {
+                            value = (TValue)Enum.ToObject(valueType, byte8);
```

Comment:
You can use `Unsafe.As` for here and all other cases as well.

Reply:
> Maybe I misunderstood - if you Unsafe.As<int, T> only when T is int32 (or uint32), then it is certainly fine.
This has endianness bug.

I thought that unsafe.as was ok only in case of coverted type equal to T  the endianess is handled correctly also if on int32(byte max value) is casted to byte?

Reply:
Right, you need to do a regular cast first to the right size, and only use Unsafe.As on the local of the right size.

Reply:
So `value = Unsafe.As<int, TValue>(ref byte8);` is lecit TValue will be byte byte8(int) has got only 1 byte(with correct endianness) valued(thanks to bound check)...only to be sure to understand all semantic.